src/components/draggable-list.tsx

```diff
@@ -0,0 +1,182 @@
+import React, { useState,useRef } from 'react';
+import { Button, Tile } from 'carbon-components-react';
+import { css, cx } from 'emotion';
+import {
+	AddAlt32,
+	Draggable16,
+	TrashCan32
+} from '@carbon/icons-react';
+
+const addDragTargetStyle = css`
+	height: var(--drag-target-height, 32px);
+	width: 100%;
+	display: flex;
+	justify-content: center;
+	outline-offset: -2px;
+	transition: height 0.15s ease-in-out;
+`;
+
+const addButtonDraggingStyle = css`
+	outline: var(--outline, 1px dashed #1666fe);
+`;
+
+const addButtonStyle = css`
+	width: 100%;
+	display: flex;
+	justify-content: center;
+`;
+
+const trashButtonStyle = css`
+	margin-left: 12px;
+	top: 0;
+	right: 0;
+	border-color: transparent;
+
+	&.bx--btn.bx--btn--icon-only.bx--tooltip__trigger {
+		position: absolute;
+	}
+`;
+
+const draggableIconStyle = css`
+	position: absolute;
+	top: 50%;
+	left: 0;
+`;
+
+const tileStyle = css`
+	position: relative;
+`;
+
+export const DraggableTileList = ({
+	template, // Functional component
+	dragOver = (_: any) => { },	// Override onDragOver event
+	dataList,
+	setDataList,
+	defaultObject // Default object created
+}: any) => {
+	const [dragging, setDragging] = useState(false);
+	const draggedItem = useRef<HTMLDivElement>();
+	const [draggedIndex, setDraggedIndex] = useState(-1);
+
+	const onDragStart = (event: any, index: number) => {
+		setDragging(true);
+		setDraggedIndex(index);
+		event.dataTransfer.setData('item-data', JSON.stringify(dataList[index]));
+		draggedItem.current = event.currentTarget;
+	};
+
+	const onDragEnd = (_: any) => {
+		setDragging(false);
+		setDraggedIndex(-1);
+		draggedItem.current = undefined;
+	};
+
+	const onDragOver = (event: any) => {
+		// Prevent drop if user
+		if(dragOver(event) === false) {
+			return false;
+		}
+
+		event.preventDefault();
+		// Adds styles only if dragged item reference exists
+		// otherwise enters height adjustment loop
+		if(draggedItem.current?.clientHeight) {
+			event.currentTarget.style.setProperty('--drag-target-height', `${(draggedItem.current?.clientHeight)}px`);
+			event.currentTarget.style.setProperty('--outline', '2px dashed #0f62fe');
+		}
+	};
+
+	const onDragLeave = (event: any) => {
+		event.currentTarget.style.setProperty('--drag-target-height', '');
+		event.currentTarget.style.setProperty('--outline', '');
+	};
+
+	const onDrop = (event: any, index: number) => {
+		event.preventDefault();
+
+		// parse only if data exists
+		const data = event.dataTransfer.getData('item-data');
+		if(!data) {
+			return false;
+		}
+		const item = JSON.parse(data);
+		const newList = [...dataList];
+
+		// Splice makes it easier to remove & add to new position
+		// Only remove item from list if item is part of list
+		if(draggedIndex !== -1) {
```

Comment:
```suggestion
		if (draggedIndex >= 0) {
```
would be a more defensive way to do it.

Also, we add a space after `if` throughout the codebase so perhaps add that to lint options